Automatically archive emails created more than 6 months ago

equivalentideas opened the pull request:
This PR creates a daily cronjob that runs a new rake task which archives all emails older than 6 months to prevent the cuttlefish database from consuming too much disk space.

Fixes https://github.com/mlandauer/cuttlefish/issues/232

equivalentideas commented:
@henare I successfully provisioned this locally, but haven't tested the actual archiving happens, relying on knowing that the other archiving task works.

I've tests the query for collecting the correct date range on production using the rails console.

Could you please have a squizo at this?

henare commented on lib/tasks/cuttlefish.rake:
```diff
@@ -16,6 +16,18 @@ namespace :cuttlefish do
   end
 
+  desc "Archive all emails created more than 6 months ago"
+  task :archive_6_month_old_emails => :environment do
```
I'd just call this something like `autoarchive`. That way if we decide to change it to say, 3 months, then we don't need to change job names everywhere.

henare commented on lib/tasks/cuttlefish.rake:
```diff
@@ -16,6 +16,18 @@ namespace :cuttlefish do
+  desc "Archive all emails created more than 6 months ago"
+  task :archive_6_month_old_emails => :environment do
+    date_of_oldest_email = Delivery.order(:created_at).first.created_at.to_date
+    if date_of_oldest_email < 6.months.ago.to_date
```
`6.month.ago` is repeated and you can DRY this up. Reducing the duplication will make it clearer, less error-prone, and easier to modify in the future.

henare commented on lib/tasks/cuttlefish.rake:
```diff
@@ -16,6 +16,18 @@ namespace :cuttlefish do
+    date_of_oldest_email = Delivery.order(:created_at).first.created_at.to_date
+    if date_of_oldest_email < 6.months.ago.to_date
+      (date_of_oldest_email...6.months.ago.to_date).each do |date|
+        Archiving.archive(date)
```
This method needs to be updated to fulfil [this requirement](https://github.com/mlandauer/cuttlefish/issues/232#issuecomment-280497491):

>This should also include removing the archive files from disk after they're transferred to S3, which is something we do manually at the moment.

henare commented on provisioning/roles/cuttlefish-app/tasks/main.yml:
```diff
@@ -92,3 +92,11 @@
     mode: 0744
+
+- name: Create old email archiving cronjob
+  copy:
```
We should use the Ansible [`cron` module](https://docs.ansible.com/ansible/cron_module.html) for this. Here's an example of a [rake task being set up using this module](https://github.com/openaustralia/morph/blob/3a5e7c5693e0811a59246dc869b9692d04524b56/provisioning/roles/morph-app/tasks/main.yml#L132-L133) for morph.

equivalentideas replied:
The reason I used the copy module instead of the cron module is that I couldn't verify that it had actually created the job. I couldn't work out where it was making the addition. I suspect a closer reading of the module docs will reveal this :)

henare replied:
It's being added to the _user crontab_ so to check it you use the `crontab` command:

```
# List the current user's crontab
crontab -l
# Edit the current user's crontab
crontab -e
```

This [section of Ubuntu's CronHowto](https://help.ubuntu.com/community/CronHowto#Two_Other_Types_of_Crontab) should give you an idea of the differences between this and the files in `/etc/cron*`.

henare commented on provisioning/roles/cuttlefish-app/tasks/main.yml:
```diff
@@ -92,3 +92,11 @@
+- name: Create old email archiving cronjob
+  copy:
+    src: archiving
+    dest: /etc/cron.daily/archiving
```
When will this run? Is it important that it runs at a particular time, e.g. when the server isn't heavily loaded?

equivalentideas replied:
@henare I wasn't sure about this. Is there a specific time of day that cuttlefish is doing less work? I should look at the linode charts to see!

henare replied:
That's a good idea. I'm genuinely not sure.

Another thing to check is if there's other scheduled tasks that run at the same time (especially things like backups). I also double-checked that if you have multiple jobs in `cron.daily` they're run sequentially, not concurrently.

equivalentideas commented:
TODOs:

* [ ] [change name of the action](https://github.com/mlandauer/cuttlefish/pull/270#discussion_r102105940)
* [ ] (Remove duplicate production of time to archive till)[https://github.com/mlandauer/cuttlefish/pull/270#discussion_r102106064]
* [ ] (Remove backups after they're sent to S3)[https://github.com/mlandauer/cuttlefish/pull/270#discussion_r102106147] (also check they they've been sent correctly?)
* [ ] (Use the ansible cron module, not copy)[https://github.com/mlandauer/cuttlefish/pull/270#discussion_r102107035]
* [ ] (Check when a good time to run this is and run it at that time)[https://github.com/mlandauer/cuttlefish/pull/270#discussion_r102107119]

equivalentideas commented:
I'm going to rebase this on top of master, cause I want to new local deploy stuff to test this on my local box.

equivalentideas commented:
> I'm going to rebase this on top of master, cause I want to new local deploy stuff to test this on my local box.

I got confused. We don't actually want that deploy stuff on this branch, because it only relates to the openaustralia fork, not upstream.

I'll just merge a version of it locally to test out deploying.

equivalentideas commented:
@henare this is ready for another review now. I've worked through those points and added tests for the new behaviour as well (and made a small refactor to extract some shared test setup).

henare commented on lib/archiving.rb:
```diff
@@ -27,11 +27,16 @@ def self.archive(date)
       copy_to_s3(date)
+
+      puts "Removing temp local file #{date}.tar.gz copied to S3..."
+      File.delete("db/archive/#{date}.tar.gz")
```
This filename needs to be DRYed up.

henare replied:
Am I right in thinking that if you don't have S3 configured, or there's a problem with it, that this archive method will actually just irretrievably delete stuff from the database? That doesn't sound like a good thing and definitely not what you'd expect a method called "archive" to do!

equivalentideas replied:
Yep that's real bad. Good catch @henare .

I'll need to add tests for the copy_to_s3 method and the archive method to make sure it does what we want. I find it too scary without tests for the fail cases now that we're deleting stuff.

equivalentideas replied:
I've now changed this so that it if S3 isn't configured, or there's an error raised when it tries to upload, it won't delete the file and tells you what it's doing on the console output.

There's tests covering this here: https://github.com/mlandauer/cuttlefish/pull/270/files#diff-01d06a85c92ee1298228bf80d6880c96R49

henare commented on provisioning/roles/cuttlefish-app/tasks/main.yml:
```diff
@@ -92,3 +92,11 @@
     group: root
     mode: 0744
+
+- name: Create old email archiving daily cronjob for 1am
```
Why 1 am?

equivalentideas replied:
> I've set this to run at 1am, which is an hour after the network and
CPU intensive backup job.

From the charts in linode, nighttime is a quieter than in the day. My thinking was that night time is a good time to run this then, so that if something goes wrong or the app crashes, not as many people will see it and we can get to it first thing. The backup runs at midnight and consumes heaps of memory for about 15/20 min. On the next hour is quiet and seemed like a good idea to do it before it's too late so someone could stay up or might be awake if it needs attention.

henare replied:
Nice one! Good analysis :+1: 

You might like to double-check that this is going to run at the time you expect. Since we're not the only ones using this server I'm pretty sure it's set to UTC, not AEST. Does this mean the job will run at 01:00 UTC? Is that the time that matches up to your analysis?

equivalentideas replied:
> You might like to double-check that this is going to run at the time you expect.

Lol good catch. It is set to UTC. So it isn't running when I thought. Since it's all far from the backup job, I recon 11pm local (23:00 local = 12:00 utc) would be a good time to run this then since it has the same benefits and there's a better chance someone will be awake.

equivalentideas replied:
Here's the graphs by the way https://manager.linode.com/linodes/graphs/cuttlefish?date=daily

equivalentideas replied:
I've made that change here https://github.com/mlandauer/cuttlefish/pull/270/commits/367e9238713cbd5943b141a2738a496433a10fcb

Is the note about AEST too specific to the OAF implementation for upstream?

henare commented on spec/lib/archiving_spec.rb:
```diff
@@ -76,4 +84,31 @@
+    end
+
+    it "removes the temp archive file it creates" do
+      delivery = Delivery.create!(
```
Unused variable assignment.

henare commented on spec/lib/archiving_spec.rb:
```diff
@@ -76,4 +84,31 @@
+
+    it "removes the temp archive file it creates" do
+      delivery = Delivery.create!(
+        id: 5,
```
Do we really need to set all of these specific attributes for this test? Would this be better created by FactoryGirl anyway?

equivalentideas commented:
TODOs:

* [ ] Refactor tests
* [ ] Add proper test coverage for Archiving.archive
* [ ] Dry up file name setting in Archiving.archive
* [ ] Make sure we never loose the archive file, e.g. if it doesn't get sent to S3 successfully, don't dump it.

equivalentideas commented:
@henare I've made those changes.

* added tests for Archiving.copy_to_s3, Archiving.archive and Archiving.unarchive so that I could change the behaviour and refactor. This involved adding vcr and climate control gems for use in tests.
* archive filename and directory are just set in one place now.
* removed duplication from the tests and set them up to use factories.
* [changed the behaviour to not delete the local archive file if s3 isn't configured or something goes wrong with the upload.](https://github.com/mlandauer/cuttlefish/pull/270#discussion_r106066725)
* [changed the cronjob to run at 12:00 UTC which matches the time I actually want it to run now](https://github.com/mlandauer/cuttlefish/pull/270#discussion_r103376601)

Would you mind doing another review for me when you get a chance?
